kopf/k8s/classes.py

```diff
@@ -0,0 +1,48 @@
+import json
+from typing import Type
+
+import pykube
+
+from kopf.k8s import config
+
+
+# TODO: this mixin has to be migrated into pykube-ng itself.
+class APIObject(pykube.objects.APIObject):
+    def patch(self, patch):
+        '''
+        Patch the Kubernetes resource by calling the API.
+        '''
+        r = self.api.patch(**self.api_kwargs(
+            headers={"Content-Type": "application/merge-patch+json"},
+            data=json.dumps(patch),
+        ))
+        self.api.raise_for_status(r)
+        self.set_obj(r.json())
+
+
+class NamespacedAPIObject(pykube.objects.NamespacedAPIObject, APIObject):
+    pass
+
+
+class CustomResourceDefinition(APIObject):
+    version = "apiextensions.k8s.io/v1beta1"
+    endpoint = "customresourcedefinitions"
+    kind = "CustomResourceDefinition"
+
+
+def _make_cls(resource) -> Type[APIObject]:
```

Comment:
I guess this can be part of pykube-ng, too

Reply:
@hjacobs In theory, there is already an implementation in `pykube.object_factory()`, and it is an officially documented way of creating the classes.

But I had issues with it. First, the self-made mixins with the `patch()` method (now solved). Second, because it requires `kind` as a search criterion, and Kopf uses the canonical plural names from the `name` field (as in the K8s API URLs, and as in the original Kubernetes client library).

The automatic implicit mapping kubectl-style (i.e. any form is accepted, including the short-names) is planned in #57, and it will be more convenient for the users, but it is not yet there.

I would be nice if pukube-ng could search by any name field (kind, plural, singular, all shortnames) — same as kubectl does. But it can be done as a separate PR in Kopf.